_posts/2021-11-18-performance-evaluation-of-timeseries.md

```diff
@@ -0,0 +1,472 @@
+
+But we’re yet to discuss the most interesting part. Time series is primarily used for quick aggregate counting. Let’s
+see what the comparison looks like when calculating the arithmetic mean in a given time interval.
+
+The script below (saved as `gen-aggregate.sh`) creates a list of queries calculating the arithmetic mean of ratings for
+a randomly selected six-hour interval:
+
+```shell
+#!/bin/bash
+RANDOM=42
+for ((i = 1; i <= $1; i++ ));
+do
+  x1=$(( $RANDOM % 10000000))
+  x2=$(( x1 + 360))
+  t1=$(date -jv +${x1}M -f "%Y-%m-%d %H:%M:%S" "2021-10-22 0:00:00" +%Y-%m-%dT%H:%M:%S)
+  t2=$(date -jv +${x2}M -f "%Y-%m-%d %H:%M:%S" "2021-10-22 0:00:00" +%Y-%m-%dT%H:%M:%S)
+  echo "db.getCollection('$2').aggregate([{ \$match: { "timestamp" : " \
+    "{\$gte:new ISODate('$t1'),\$lt:new ISODate('$t2')} } }," \
+    "{ \$group: { _id: null, avg: { \$avg: \"\$rating\" } } }])"
+done >> aggregate-$2-$1.js
+```
+
+I prepared three script sets with: 10K, 50K and 100K queries for both collections:
+
+```shell
+./gen-aggregate.sh 10000 coll-ts
+./gen-aggregate.sh 50000 coll-ts
+./gen-aggregate.sh 100000 coll-ts
+
+./gen-aggregate.sh 10000 coll-ord
+./gen-aggregate.sh 50000 coll-ord
+./gen-aggregate.sh 100000 coll-ord
+```
+
+I made the measurements using following commands:
+
+```shell
+time mongo test aggregate-coll-ts-10000.js
+time mongo test aggregate-coll-ord-10000.js
+
+time mongo test aggregate-coll-ts-50000.js
+time mongo test aggregate-coll-ord-50000.js
+
+time mongo test aggregate-coll-ts-100000.js
+time mongo test aggregate-coll-ord-100000.js
+
+```
+
+The results are shown in the table below:
+
+| Number of queries | Normal collection [min:sec] | Time series [min:sec] | Diff |
+| :----: | ----: | ----: | :----: |
+| 10000 | 0:21,947 | 0:16,835 | -23% |
+| 50000 | 1:37,02 | 1:11,18 | -26% |
+| 100000 | 4:33,29 | 2:21,37 | -48% |
+
+Although there are far fewer queries this time than in the previous experiment, the differences in times are much
+greater. It clearly proves that time series collections are indeed spreading their wings when we want to use aggregation
+queries and have been developed mainly for this purpose. The reason is probably the sequential way of writing data,
+which shows a noticeable improvement when running ranged queries. The results clearly show also how much of an impact
+query performance can have on an element that often doesn’t get the adequate attention: proper modelling the data.
+
+## Limitations
+
+Of course, time series collections also have their limitations and should not be used as a golden hammer.
+We have already mentioned the first one — the lack of the primary key index. It stems from the assumption
+that searches will be primarily based on time, so there is no point in creating an index that will be useful for very
+few people. Of course, we can create this index ourselves.
+
+It is also not possible to sort by time field, which is another inconvenience. If we want to have
```

Comment:
It is also not possible to sort by time field, ... -> It is not possible to sort by time field, either, ...